packages/contract/src/contract-factory.ts

```diff
@@ -146,26 +155,188 @@ export default class ContractFactory {
+        throw new FuelError(
+          ErrorCode.MAX_FEE_TOO_LOW,
+          `Max fee '${options.maxFee}' is lower than the required: '${txCost.maxFee}'.`
+        );
+      }
+    } else {
+      request.maxFee = txCost.maxFee;
+    }
+
+    await account.fund(request, txCost);
+
+    return request;
+  }
+
+  /**
+   * Deploy a contract of any length with the specified options.
    *
    * @param deployOptions - Options for deploying the contract.
    * @returns A promise that resolves to the deployed contract instance.
    */
   async deploy<TContract extends Contract = Contract>(
     deployOptions: DeployContractOptions = {}
   ): Promise<DeployContractResult<TContract>> {
-    if (this.bytecode.length > MAX_CONTRACT_SIZE) {
+    const account = this.getAccount();
+    const { consensusParameters } = account.provider.getChain();
+    const maxContractSize = consensusParameters.contractParameters.contractMaxSize.toNumber();
+
+    if (this.bytecode.length > maxContractSize) {
+      return this.deployContractAsBlobs(deployOptions);
+    }
+    return this.deployContract(deployOptions);
+  }
+
+  /**
+   * Deploy a contract with the specified options.
+   *
+   * @param deployOptions - Options for deploying the contract.
+   * @returns A promise that resolves to the deployed contract instance.
+   */
+  async deployContract<TContract extends Contract = Contract>(
+    deployOptions: DeployContractOptions = {}
+  ): Promise<DeployContractResult<TContract>> {
+    const account = this.getAccount();
+    const { consensusParameters } = account.provider.getChain();
+    const maxContractSize = consensusParameters.contractParameters.contractMaxSize.toNumber();
+
+    if (this.bytecode.length > maxContractSize) {
       throw new FuelError(
         ErrorCode.CONTRACT_SIZE_EXCEEDS_LIMIT,
-        'Contract bytecode is too large. Max contract size is 100KB'
+        'Contract bytecode is too large. Max contract size is 100KB' // change error
       );
     }
 
     const { contractId, transactionRequest } = await this.prepareDeploy(deployOptions);
 
+    const transactionResponse = await account.sendTransaction(transactionRequest);
+
+    const waitForResult = async () => {
+      const transactionResult = await transactionResponse.waitForResult<TransactionType.Create>();
+      const contract = new Contract(contractId, this.interface, account) as TContract;
+
+      return { contract, transactionResult };
+    };
+
+    return { waitForResult, contractId, transactionId: transactionResponse.id };
+  }
+
+  /**
+   * Chunks and deploys a contract via a loader contract. Suitable for deploying contracts larger than the max contract size.
+   *
+   * @param deployOptions - Options for deploying the contract.
+   * @returns A promise that resolves to the deployed contract instance.
+   */
+  async deployContractAsBlobs<TContract extends Contract = Contract>(
+    deployOptions: DeployContractOptions = {
+      chunkSizeTolerance: CHUNK_SIZE_TOLERANCE,
```

Comment:
@danielbate pardon my ignorance, but I'm having difficulty determining how I could derive an appropriate `chunkSizeTolerance` programmatically, if the is tied to fee fluctuations. Perhaps I am not understanding the context of how to properly estimate that.

Would it be possible to determine the optimal chunk size based on the account's current balance? I wonder if we could iteratively determine a cost for the user and thus avoid the need for a user provided value and also increase the probability that we only deploy in chunk sizes that will fall within the available balance.

@Torres-ssf made a great recommendation [here](https://github.com/FuelLabs/fuels-ts/pull/2827#discussion_r1707915203) which made me think it seems possible.

```suggestion
  const chainInfo = account.provider.getChain();
  const gasPrice = await account.provider.estimateGasPrice(10);

  // Function to estimate cost for a given chunk size
  const estimateCost = (chunkSize: number) => {
    const chunks = getContractChunks(this.bytecode, chunkSize);
    let totalCost = bn(0);
    const estimatedBlobIds: string[] = [];

    for (const chunk of chunks) {
      const transactionRequest = this.blobTransactionRequest({ bytecode: chunk.bytecode });
      if (!estimatedBlobIds.includes(transactionRequest.blobId)) {
        const minGas = transactionRequest.calculateMinGas(chainInfo);
        const minFee = calculateGasFee({
          gasPrice,
          gas: minGas,
          priceFactor: chainInfo.consensusParameters.feeParameters.gasPriceFactor,
          tip: transactionRequest.tip,
        }).add(1);
        totalCost = totalCost.add(minFee);
        estimatedBlobIds.push(transactionRequest.blobId);
      }
    }

    return { totalCost, chunkCount: chunks.length };
  };

  //search for optimal chunk size
  let minSize = WORD_SIZE; // Minimum chunk size
  let maxSize = this.getMaxChunkSize(); // Maximum possible chunk size
  let optimalSize = maxSize;
  let optimalCost = estimateCost(maxSize);

  while (minSize <= maxSize) {
    const midSize = Math.floor((minSize + maxSize) / 2);
    const { totalCost, chunkCount } = estimateCost(midSize);

    if (totalCost.lte(await account.getBalance())) {
      optimalSize = midSize;
      optimalCost = { totalCost, chunkCount };
      minSize = midSize + 1;
    } else {
      maxSize = midSize - 1;
    }
  }
```

Thoughts? @Torres-ssf @danielbate

Reply:
No good suggestion mate, I think there are improvements that can be made to the chunk optimisation tbh, potentially as follow up, but I'll take a look at this today. @Torres-ssf suggestion has been added 👍🏻 

Additionally, re chunk size tolerance, there are some parameters that are out of our control when coming to the chunk size at the time of estimation.  I should've also mentioned that the contract size consensus param isn't necessarily the upper limit. We also have a request size limit for HTTP requests against the server. Contract size is configurable by the chain and therefore could be set higher than the request size limit. Therefore we need some amount of user configurable tolerance for unknowns at the time of estimation.

Reply:
Thanks for the clarification @danielbate 

I'm definitely open to this being a follow-up for sure, as I think the current implementation satisfies the current requirements of getting larger contracts deployed. 

I am still wrapping my mind around how one would derive a reasonable tolerance though, based on what you're saying, my understanding is that the upper limit is the max of the request size and the contract size, but shouldn't both of those parameters be retrievable at compile time? 

Perhaps we can have a call to discuss.

Reply:
So we have the transaction limit and contract limit, both retrievable via the consensus params (I actually hadn't accounted for tx limit so thats's now been added). However graphQL also has a maximum configurable request body size to contend with which will also return a `413` which I'm not sure we can retrieve.

We've also experienced some flakiness in devnet and the manual override has been really helpful.

But it's definitely worth the call.

Reply:
After discussing with @danielbate we decided we will plan for this in the future

https://github.com/FuelLabs/fuels-ts/issues/2917